src/OrchardCore/OrchardCore.Users.Core/Models/User.cs

```diff
@@ -15,12 +15,12 @@ public class User : Entity, IUser
         public string SecurityStamp { get; set; }
         public bool EmailConfirmed { get; set; }
         public string ResetToken { get; set; }
-        public IList<string> RoleNames { get; set; } = new List<string>();
-        public IList<UserClaim> UserClaims { get; set; } = new List<UserClaim>();
-        public IList<UserLoginInfo> LoginInfos { get; set; } = new List<UserLoginInfo>();
+        public ImmutableArray<string> RoleNames { get; set; } = ImmutableArray<string>.Empty;
```

Comment:
Why? Is a User object used in concurrent threads?

Reply:
As i remember, there are some places where we mutate it and i checked that we act on a same shared instance.

Let me some time to retrieve an example

Reply:
E.g in `UserStore.AddToRoleAsync()` line 332

            ((User)user).RoleNames.Add(roleName);

That becomes here

            ((User)user).RoleNames = ((User)user).RoleNames.Add(roleName);

Reply:
But only one thread accesses this user instance, so the list is thread-safe.

Reply:
Yes normally but just tried 2 concurrent editing (by using a break point) of the same user, i could see that the 2 concurrent thread are mutating the same user instance and then its roles collection.

The other problem is when we are using / enumerating on e.g roleNames while mutating it.

Reply:
Interesting. Where are we mutating and iterating at the same time, if not on two different thread.

>  just tried 2 concurrent editing

This is worrying. Can you explain how that happens?

Reply:
`UserManager` is scoped, it resolves `UserStore` which is scoped too, which uses `Session` to load a user. In the end I don't understand how two thread would access the same user instance, unless we have some wrong parallel code.

Reply:
Okay let me check again

Reply:
Oh, just retried, you're right there are 2 different instances, i don't know what i did, maybe too tired ;) These were some of the last changes i did as for others singletons and without doing the above test, indeed there are all scoped services, i'm stupid.

Thanks, i will revert all the changes around user and roles.

Reply:
So, i don't know why i did here the same changes as for other services but using a singleton cache.

Sorry, i reverted the changes related to users and roles.
I reverted